File: apps/mobile/ios/Sources/Components/TchatInput.swift
//
//  TchatInput.swift
//  TchatApp
//
//  Created by Claude on 21/09/2024.
//

import SwiftUI

/// Input field component following Tchat design system
public struct TchatInput: View {

    // MARK: - Input Types
    public enum InputType {
        case text
        case email
        case password
        case number
        case search
        case multiline(minLines: Int = 3, maxLines: Int = 6)
    }

    public enum ValidationState {
        case none
        case valid
        case invalid(message: String)
    }

    public enum Size {
        case small
        case medium
        case large
    }

    // MARK: - Properties
    @Binding private var text: String
    @FocusState private var isFocused: Bool

    let placeholder: String
    let type: InputType
    let size: Size
    let validationState: ValidationState
    let isDisabled: Bool
    let leadingIcon: String?
    let trailingIcon: String?
    let onTrailingIconTap: (() -> Void)?

    // MARK: - Private Properties
    private let colors = Colors()
    @State private var isSecureTextVisible = false

    // MARK: - Initializer
    public init(
        text: Binding<String>,
        placeholder: String,
        type: InputType = .text,
        size: Size = .medium,
        validationState: ValidationState = .none,
        isDisabled: Bool = false,
        leadingIcon: String? = nil,
        trailingIcon: String? = nil,
        onTrailingIconTap: (() -> Void)? = nil
    ) {
        self._text = text
        self.placeholder = placeholder
        self.type = type
        self.size = size
        self.validationState = validationState
        self.isDisabled = isDisabled
        self.leadingIcon = leadingIcon
        self.trailingIcon = trailingIcon
        self.onTrailingIconTap = onTrailingIconTap
    }

    // MARK: - Body
    public var body: some View {
        VStack(alignment: .leading, spacing: Spacing.xs) {
            inputField
                .disabled(isDisabled)

            if case .invalid(let message) = validationState {
                Text(message)
                    .font(.caption)
                    .foregroundColor(colors.error)
                    .padding(.horizontal, Spacing.xs)
            }
        }
    }

    // MARK: - Input Field
    @ViewBuilder
    private var inputField: some View {
        HStack(spacing: Spacing.xs) {
            if let leadingIcon = leadingIcon {
                Image(systemName: leadingIcon)
                    .foregroundColor(iconColor)
                    .frame(width: 16, height: 16)
            }

            inputContent

            trailingIconView
        }
        .padding(.horizontal, horizontalPadding)
        .padding(.vertical, verticalPadding)
        .background(backgroundColor)
        .overlay(
            RoundedRectangle(cornerRadius: Spacing.sm)
                .stroke(borderColor, lineWidth: borderWidth)
        )
        .cornerRadius(Spacing.sm)
        .animation(.easeInOut(duration: 0.2), value: isFocused)
        .animation(.easeInOut(duration: 0.2), value: validationState)
    }

    @ViewBuilder
    private var inputContent: some View {
        switch type {
        case .text:
            TextField(placeholder, text: $text)
                .textFieldStyle(TchatTextFieldStyle())
                .focused($isFocused)

        case .email:
            TextField(placeholder, text: $text)
                .textFieldStyle(TchatTextFieldStyle())
                .keyboardType(.emailAddress)
                .autocapitalization(.none)
                .focused($isFocused)

        case .password:
            if isSecureTextVisible {
                TextField(placeholder, text: $text)
                    .textFieldStyle(TchatTextFieldStyle())
                    .focused($isFocused)
            } else {
                SecureField(placeholder, text: $text)
                    .textFieldStyle(TchatTextFieldStyle())
                    .focused($isFocused)
            }

        case .number:
            TextField(placeholder, text: $text)
                .textFieldStyle(TchatTextFieldStyle())
                .keyboardType(.numberPad)
                .focused($isFocused)

        case .search:
            TextField(placeholder, text: $text)
                .textFieldStyle(TchatTextFieldStyle())
                .autocapitalization(.none)
                .focused($isFocused)

        case .multiline(let minLines, let maxLines):
            TextEditor(text: $text)
                .frame(minHeight: CGFloat(minLines) * 20, maxHeight: CGFloat(maxLines) * 20)
                .focused($isFocused)
        }
    }

    @ViewBuilder
    private var trailingIconView: some View {
        if type == .password {
            Button(action: {
                isSecureTextVisible.toggle()
            }) {
                Image(systemName: isSecureTextVisible ? "eye.slash" : "eye")
                    .foregroundColor(iconColor)
                    .frame(width: 16, height: 16)
            }
        } else if let trailingIcon = trailingIcon {
            Button(action: {
                onTrailingIconTap?()
            }) {
                Image(systemName: trailingIcon)
                    .foregroundColor(iconColor)
                    .frame(width: 16, height: 16)
            }
        }
    }

    // MARK: - Computed Properties

    private var inputFont: Font {
        switch size {
        case .small:
            return .system(size: 14)
        case .medium:
            return .system(size: 16)
        case .large:
            return .system(size: 18)
        }
    }

    private var horizontalPadding: CGFloat {
        switch size {
        case .small: return Spacing.sm
        case .medium: return Spacing.md
        case .large: return Spacing.lg
        }
    }

    private var verticalPadding: CGFloat {
        switch size {
        case .small: return Spacing.xs
        case .medium: return Spacing.sm
        case .large: return Spacing.md
        }
    }

    private var backgroundColor: Color {
        if isDisabled {
            return colors.surface.opacity(0.5)
        }
        return colors.background
    }

    private var borderColor: Color {
        if isDisabled {
            return colors.border.opacity(0.5)
        }

        switch validationState {
        case .none:
            return isFocused ? colors.borderFocus : colors.border
        case .valid:
            return colors.success
        case .invalid:
            return colors.borderError
        }
    }

    private var borderWidth: CGFloat {
        switch validationState {
        case .none:
            return isFocused ? 2 : 1
        case .valid, .invalid:
            return 2
        }
    }

    private var iconColor: Color {
        if isDisabled {
            return colors.textDisabled
        }
        return isFocused ? colors.primary : colors.textSecondary
    }
}

// MARK: - Text Field Style
struct TchatTextFieldStyle: TextFieldStyle {
    func _body(configuration: TextField<Self._Label>) -> some View {
        configuration
            .foregroundColor(Colors().textPrimary)
    }
}

// MARK: - Preview
#if DEBUG
struct TchatInput_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: Spacing.md) {
            TchatInput(
                text: .constant(""),
                placeholder: "Enter your email",
                type: .email,
                leadingIcon: "envelope"
            )

            TchatInput(
                text: .constant(""),
                placeholder: "Enter password",
                type: .password,
                leadingIcon: "lock"
            )

            TchatInput(
                text: .constant("Search..."),
                placeholder: "Search",
                type: .search,
                leadingIcon: "magnifyingglass",
                trailingIcon: "xmark.circle.fill"
            )

            TchatInput(
                text: .constant("Valid input"),
                placeholder: "Valid input",
                validationState: .valid,
                leadingIcon: "checkmark.circle"
            )

            TchatInput(
                text: .constant("Invalid input"),
                placeholder: "Invalid input",
                validationState: .invalid(message: "This field is required"),
                leadingIcon: "exclamationmark.triangle"
            )

            TchatInput(
                text: .constant(""),
                placeholder: "Disabled input",
                isDisabled: true
            )

            TchatInput(
                text: .constant(""),
                placeholder: "Type your message here...",
                type: .multiline(minLines: 3, maxLines: 6)
            )

            HStack(spacing: Spacing.sm) {
                TchatInput(
                    text: .constant("Small"),
                    placeholder: "Small",
                    size: .small
                )
                TchatInput(
                    text: .constant("Medium"),
                    placeholder: "Medium",
                    size: .medium
                )
                TchatInput(
                    text: .constant("Large"),
                    placeholder: "Large",
                    size: .large
                )
            }
        }
        .padding()
        .background(Color(.systemBackground))
        .previewLayout(.sizeThatFits)
    }
}
#endif

File: apps/mobile/android/app/src/main/java/com/tchat/components/TchatInput.kt
package com.tchat.components

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.tchat.designsystem.Colors
import com.tchat.designsystem.Spacing

/**
 * Input field component following Tchat design system
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun TchatInput(
    value: String,
    onValueChange: (String) -> Unit,
    placeholder: String,
    modifier: Modifier = Modifier,
    type: TchatInputType = TchatInputType.Text,
    size: TchatInputSize = TchatInputSize.Medium,
    validationState: TchatValidationState = TchatValidationState.None,
    enabled: Boolean = true,
    leadingIcon: ImageVector? = null,
    trailingIcon: ImageVector? = null,
    onTrailingIconClick: (() -> Unit)? = null,
    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
    keyboardActions: KeyboardActions = KeyboardActions.Default,
    singleLine: Boolean = true,
    maxLines: Int = if (singleLine) 1 else Int.MAX_VALUE,
    minLines: Int = 1
) {
    var isFocused by remember { mutableStateOf(false) }
    var passwordVisible by remember { mutableStateOf(false) }
    val focusRequester = remember { FocusRequester() }

    // Animation values
    val borderColor by animateColorAsState(
        targetValue = when {
            !enabled -> Colors.border.copy(alpha = 0.5f)
            validationState is TchatValidationState.Invalid -> Colors.borderError
            validationState is TchatValidationState.Valid -> Colors.success
            isFocused -> Colors.borderFocus
            else -> Colors.border
        },
        label = "border_color"
    )

    val borderWidth by animateFloatAsState(
        targetValue = when {
            validationState !is TchatValidationState.None -> 2f
            isFocused -> 2f
            else -> 1f
        },
        label = "border_width"
    )

    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(Spacing.xs)
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .border(
                    BorderStroke(borderWidth.dp, borderColor),
                    RoundedCornerShape(Spacing.sm)
                )
        ) {
            OutlinedTextField(
                value = value,
                onValueChange = onValueChange,
                placeholder = {
                    Text(
                        text = placeholder,
                        fontSize = size.fontSize,
                        color = Colors.textTertiary
                    )
                },
                modifier = Modifier
                    .fillMaxWidth()
                    .focusRequester(focusRequester)
                    .onFocusChanged { isFocused = it.isFocused },
                enabled = enabled,
                singleLine = singleLine,
                maxLines = maxLines,
                minLines = minLines,
                textStyle = LocalTextStyle.current.copy(
                    fontSize = size.fontSize,
                    color = if (enabled) Colors.textPrimary else Colors.textDisabled
                ),
                keyboardOptions = keyboardOptions.copy(
                    keyboardType = type.keyboardType,
                    imeAction = keyboardOptions.imeAction
                ),
                keyboardActions = keyboardActions,
                visualTransformation = when (type) {
                    is TchatInputType.Password -> {
                        if (passwordVisible) VisualTransformation.None
                        else PasswordVisualTransformation()
                    }
                    else -> VisualTransformation.None
                },
                leadingIcon = if (leadingIcon != null) {
                    {
                        Icon(
                            imageVector = leadingIcon,
                            contentDescription = null,
                            tint = if (enabled) {
                                if (isFocused) Colors.primary else Colors.textSecondary
                            } else Colors.textDisabled,
                            modifier = Modifier.size(16.dp)
                        )
                    }
                } else null,
                trailingIcon = if (type is TchatInputType.Password || trailingIcon != null) {
                    {
                        when (type) {
                            is TchatInputType.Password -> {
                                IconButton(
                                    onClick = { passwordVisible = !passwordVisible }
                                ) {
                                    Icon(
                                        imageVector = if (passwordVisible) Icons.Default.VisibilityOff
                                        else Icons.Default.Visibility,
                                        contentDescription = if (passwordVisible) "Hide password"
                                        else "Show password",
                                        tint = if (enabled) Colors.textSecondary else Colors.textDisabled,
                                        modifier = Modifier.size(16.dp)
                                    )
                                }
                            }
                            else -> {
                                if (trailingIcon != null) {
                                    IconButton(
                                        onClick = { onTrailingIconClick?.invoke() }
                                    ) {
                                        Icon(
                                            imageVector = trailingIcon,
                                            contentDescription = null,
                                            tint = if (enabled) Colors.textSecondary else Colors.textDisabled,
                                            modifier = Modifier.size(16.dp)
                                        )
                                    }
                                }
                            }
                        }
                    }
                } else null,
                shape = RoundedCornerShape(Spacing.sm),
                colors = OutlinedTextFieldDefaults.colors(
                    focusedBorderColor = Color.Transparent,
                    unfocusedBorderColor = Color.Transparent,
                    disabledBorderColor = Color.Transparent,
                    errorBorderColor = Color.Transparent,
                    focusedContainerColor = if (enabled) Colors.background else Colors.surface.copy(alpha = 0.5f),
                    unfocusedContainerColor = if (enabled) Colors.background else Colors.surface.copy(alpha = 0.5f),
                    disabledContainerColor = Colors.surface.copy(alpha = 0.5f)
                )
            )
        }

        // Validation message
        if (validationState is TchatValidationState.Invalid) {
            Text(
                text = validationState.message,
                fontSize = 12.sp,
                color = Colors.error,
                modifier = Modifier.padding(horizontal = Spacing.xs)
            )
        }
    }
}

/**
 * Input type definitions
 */
sealed class TchatInputType(val keyboardType: KeyboardType) {
    object Text : TchatInputType(KeyboardType.Text)
    object Email : TchatInputType(KeyboardType.Email)
    object Password : TchatInputType(KeyboardType.Password)
    object Number : TchatInputType(KeyboardType.Number)
    object Search : TchatInputType(KeyboardType.Text)
    object Multiline : TchatInputType(KeyboardType.Text)
}

/**
 * Input size definitions
 */
enum class TchatInputSize(
    val fontSize: androidx.compose.ui.unit.TextUnit,
    val horizontalPadding: androidx.compose.ui.unit.Dp,
    val verticalPadding: androidx.compose.ui.unit.Dp
) {
    Small(
        fontSize = 14.sp,
        horizontalPadding = Spacing.sm,
        verticalPadding = Spacing.xs
    ),
    Medium(
        fontSize = 16.sp,
        horizontalPadding = Spacing.md,
        verticalPadding = Spacing.sm
    ),
    Large(
        fontSize = 18.sp,
        horizontalPadding = Spacing.lg,
        verticalPadding = Spacing.md
    )
}

/**
 * Validation state definitions
 */
sealed class TchatValidationState {
    object None : TchatValidationState()
    object Valid : TchatValidationState()
    data class Invalid(val message: String) : TchatValidationState()
}

// Preview
@Preview(showBackground = true)
@Composable
fun TchatInputPreview() {
    Column(
        modifier = Modifier.padding(Spacing.md),
        verticalArrangement = Arrangement.spacedBy(Spacing.md)
    ) {
        TchatInput(
            value = "",
            onValueChange = { },
            placeholder = "Enter your email",
            type = TchatInputType.Email,
            leadingIcon = Icons.Default.Email
        )

        TchatInput(
            value = "",
            onValueChange = { },
            placeholder = "Enter password",
            type = TchatInputType.Password,
            leadingIcon = Icons.Default.Lock
        )

        TchatInput(
            value = "Search...",
            onValueChange = { },
            placeholder = "Search",
            type = TchatInputType.Search,
            leadingIcon = Icons.Default.Search,
            trailingIcon = Icons.Default.Clear,
            onTrailingIconClick = { }
        )

        TchatInput(
            value = "Valid input",
            onValueChange = { },
            placeholder = "Valid input",
            validationState = TchatValidationState.Valid,
            leadingIcon = Icons.Default.CheckCircle
        )

        TchatInput(
            value = "Invalid input",
            onValueChange = { },
            placeholder = "Invalid input",
            validationState = TchatValidationState.Invalid("This field is required"),
            leadingIcon = Icons.Default.Warning
        )

        TchatInput(
            value = "",
            onValueChange = { },
            placeholder = "Disabled input",
            enabled = false
        )

        TchatInput(
            value = "",
            onValueChange = { },
            placeholder = "Type your message here...",
            type = TchatInputType.Multiline,
            singleLine = false,
            maxLines = 6,
            minLines = 3
        )

        Row(
            horizontalArrangement = Arrangement.spacedBy(Spacing.sm)
        ) {
            TchatInput(
                value = "Small",
                onValueChange = { },
                placeholder = "Small",
                size = TchatInputSize.Small,
                modifier = Modifier.weight(1f)
            )
            TchatInput(
                value = "Medium",
                onValueChange = { },
                placeholder = "Medium",
                size = TchatInputSize.Medium,
                modifier = Modifier.weight(1f)
            )
            TchatInput(
                value = "Large",
                onValueChange = { },
                placeholder = "Large",
                size = TchatInputSize.Large,
                modifier = Modifier.weight(1f)
            )
        }
    }
}